app/src/middleware/apiTracker.js

```diff
@@ -0,0 +1,185 @@
+ *
+ * Log statistics for CDOGS Requests.
+ *
+ * @see morgan
+ *
+ * @exports initializeApiTracker
+ */
+const moment = require('moment');
+const morgan = require('morgan');
+
+// this will suppress a console warning about moment deprecating a default fallback on non ISO/RFC2822 date formats
+// we will just force it to use the new Date constructor.
+moment.createFromInputFallback = function (config) {
+  config._d = new Date(config._i);
+};
+
+const docGenUrl = '/api/v1/docGen';
+const trackerUrls = [docGenUrl];
+
+// add in any token (custom or morgan built-in) we want to the format, then morgan can parse out later
+// status and response-time is a built-in morgan token
+const apiTrackerFormat = ':operation :azp :timestamp :contextKeyCount :contentFileType :contentEncodingType :contentSize :outputFileType :res[content-length] :status :response-time';
+
+const apiTracker = async (req, res, next) => {
+
+  if (trackerUrls.includes(req.url)) {
+    req._timestamp = moment.utc().valueOf();
+    req._operation = req.url === docGenUrl ? 'DOCGEN' : 'Unknown';
+
+    /*
+    When/If we need to parse data out of the response, we would do it here...
+    const defaultEnd = res.end;
+    const chunks = [];
+    res.end = (...restArgs) => {
+        try {
+            if (restArgs[0]) {
+                chunks.push(Buffer.from(restArgs[0]));
+            }
+            const body = Buffer.concat(chunks).toString('utf8');
+            const obj = JSON.parse(body);
+
+        } catch (err) {
+            log.error('mailApiTracker', err);
+        }
+        defaultEnd.apply(res, restArgs);
+    };
+    */
+  }
+  next();
+};
+
+const initializeApiTracker = (app) => {
+
+  // register token parser functions.
+  // this one would depend on authorizedParty middleware being loaded
+  morgan.token('azp', (req) => {
+    return req.authorizedParty ? req.authorizedParty : '-';
+  });
+
+  morgan.token('operation', (req) => {
+    return req._operation ? req._operation : '-';
+  });
+
+  morgan.token('timestamp', (req) => {
+    return req._timestamp ? req._timestamp : '-';
+  });
+
+  morgan.token('contextKeyCount', (req) => {
+    function countKeys(source) {
+      if (!source) return 0;
+      let result = 0;
+      (function count(obj) {
+        if (Array.isArray(obj)) {
+          obj.forEach(function (j) {
+            count(j);
+          });
+        } else {
+          Object.keys(obj).forEach(function (k) {
+            result++;
+            const v = obj[k];
+            if (typeof v === 'object') {
+              count(v);
+            }
+          });
+        }
+      })(source);
+      return result;
+    }
+
+    try {
+      // want to return some idea of the size/magnitude/complexity of the contexts
+      return countKeys(req.body.contexts);
+    } catch (e) {
+      return '0';
+    }
+  });
+
+  morgan.token('contentFileType', (req) => {
+    try {
+      return req.body.template.contentFileType;
+    } catch (e) {
+      return '-';
```

Comment:
Wondering if its' better to have the exception case return a null object instead of a dash... don't know which would work better for representing "unknown" values.

Reply:
it's getting put into a morgan log line, so would have to handle the null one way or another.  it's more of a morgan convention.  always want something in the log line to parse out.